### PR TITLE
feat: rewrite steering + per-bullet accept/reject/edit review (#210, #211)

### LINKED ISSUE #210: Rewrite: freeform instructions box + page-length presets (steer the rewrite prompt)
## Problem

The rewrite prompt is fully hardcoded (`buildSectionSystemPrompt()` in `rewrite-section.ts:75`, `buildSummarySystemPrompt()` in `rewrite-summary.ts`). The user cannot tell the rewriter **what they want** — e.g. "target a staff role", "emphasize leadership", "keep it concise", "lean technical". Recruidea exposes exactly this as a freeform instructions box.

Recruidea also lets the user pick a **target length (1 / 2 / 3 pages)**. Because length is the single biggest "what am I looking for" lever and maps cleanly onto a compression instruction, it is folded into this issue as a preset alongside the freeform box.

## What to borrow from Recruidea

Recruidea's pattern (`compose.tsx:649` UI → `custom_instructions` payload → `generate-outreach/index.ts:317` backend):

- A multiline textarea, `placeholder` showing example asks.
- The freeform text is **NOT** templated into the prompt body. It is appended as a **system-message suffix**:
  ```ts
  const systemSuffix = userInstructions
    ? ` The user has these additional instructions: ${userInstructions}`
    : "";
  ```

The suffix approach keeps the base prompt's number-preservation / no-fabrication guardrails intact and just layers user intent on top.

## Design decision — page length (settled)

ResumeLint's rewrite is **text-only and section-by-section** — WebLLM never sees PDF layout or pagination, so "1 page" cannot be *enforced* as true pagination. It is approximated as a **per-page length budget** injected into the system prompt:

- A page target maps to a concrete length budget (e.g. 1pg → tighter per-bullet word cap + fewer bullets/role; 2pg → moderate; 3pg → loose). Exact numbers tuned during implementation.
- Plus explicit **recency-weighted guidance**: instruct the model to **compress or combine older experience entries** preferentially, so the budget is spent on recent, relevant roles rather than trimming uniformly.

**UI shape:** page targets are **preset chips inside the freeform box** (one input, not a separate selector). Clicking a chip (e.g. `1 page` / `2 pages` / `3 pages`) prefills editable guidance text into the box and sets the target that drives the budget injection. Default = no chip selected → no length constraint → identical output to today. The user can still type freeform text alongside or instead of a chip.

## Scope

1. **Input** — a freeform "Instructions for the rewrite (optional)" textarea near the rewrite trigger (whole-resume in `ResumeRewrite.tsx`; consider per-role in `SectionRewrite.tsx` too), with **page-length preset chips** (`1 page` / `2 pages` / `3 pages`) above/inside it. Reuse a design-system input/`EditableField`-style primitive and the `Button` primitive for chips — no raw `<textarea>` / `<button>`. Example placeholder, e.g. *"e.g. 'target a staff engineer role' or 'keep bullets under 20 words'"*.
2. **Plumb** the value(s) through `useResumeRewrite` → `rewriteResumeWithLlm` → `rewriteSectionWithLlm` / `rewriteSummaryWithLlm` as an optional `userInstructions?: string` and an optional `pageTarget?: 1 | 2 | 3`.
3. **Inject** as a system-prompt suffix inside `buildSectionSystemPrompt()` / `buildSummarySystemPrompt()` — appended after the existing guardrail text, not interleaved:
   - `userInstructions` → verbatim suffix (empty string when blank → no behavior change).
   - `pageTarget` → a derived length-budget sentence **plus** the recency-compression guidance ("compress or combine older experience entries to fit"). Empty when unset.
4. Persist last-used instructions + page target in `localStorage` (`rl_*` key) so a re-run keeps them (optional, nice-to-have).

## Out of scope

- **True pagination / layout-aware length.** WebLLM sees text only; this issue approximates length via a compression budget, not a rendered page count.
- **JD-aware rewriting** — the JD-match modules already exist (`src/lib/jd-match/`) but wiring JD into rewrite is a separate concern. This issue is the **user's own freeform commentary + length target** only.

## Acceptance

- Textarea present near the rewrite trigger; blank + no chip by default → identical output to today.
- Non-empty instructions reach the LLM as a system-prompt suffix and visibly steer the rewrite.
- Selecting a page-length chip injects a length-budget + older-experience-compression instruction and visibly shortens/lengthens the rewrite toward the target.
- Page chips prefill editable text into the freeform box (one input); user can edit or clear.
- Guardrails (number preservation, no fabrication) still enforced.
- Reuses design-system primitives (no raw interactive HTML); passes style guard.

## References

- Recruidea: `dashboard/app/sponsor/outreach/compose.tsx:649` (textarea), `supabase/functions/generate-outreach/index.ts:317` (system-suffix injection)
- ResumeLint targets: `src/components/features/ResumeRewrite.tsx`, `src/hooks/useResumeRewrite.ts`, `src/lib/webllm/rewrite-section.ts:75`, `src/lib/webllm/rewrite-summary.ts`

### LINKED ISSUE #211: Rewrite: per-bullet accept/reject/edit + apply into reconstructed resume
## Problem

After a rewrite, the user can only **"copy all bullets to clipboard"** or **"Discard"** the whole proposal (`SectionRewrite.tsx:309`, `ResumeRewriteProposed.tsx:60`). There is no way to **accept or reject changes one-by-one, per-section, or all-at-once**, and no way to apply accepted changes back into the reconstructed resume. Recruidea has all three.

## Design decision (settled)

Granularity = **per-bullet, with alignment**. ResumeLint's WebLLM rewrite emits a *new* bullet list (M may ≠ N original), so there is no inherent 1:1 mapping. We greedily align each proposed bullet to its nearest original (LCS / similarity) so each proposed bullet renders as an inline diff against its matched original and is individually acceptable. Unmatched proposed bullets = pure additions; unmatched originals = pure removals.

> Cloud-side persistence/versioning of accepted edits (Recruidea's Supabase `artifact_suggestions` path) is **punted** — that is the paid-tier lane and is not filed in this public repo.

## What to borrow from Recruidea

In-memory decision model from `useArtifactReview` (drop the Supabase half):

```ts
const [decisions, setDecisions] = useState<Map<string, "accepted" | "rejected">>(new Map());
const [editedValues, setEditedValues] = useState<Map<string, string>>(new Map());
// per-change: acceptSuggestion / rejectSuggestion / editSuggestionValue (edit auto-accepts)
// batch:      acceptAllHighConfidence / rejectRemaining / resetAllDecisions
// apply:      applyAccepted -> applySuggestionsToResume(parsed, accepted, edits)  (pure)
```

## Scope

1. **Bullet alignment** helper in `src/lib/` — `alignBullets(original[], proposed[]): AlignedPair[]` where each pair is `{original?, proposed?, id}`. Pure + unit tested.
2. **Decision hook** `src/hooks/useRewriteReview.ts` — `Map<bulletId, "accepted"|"rejected">` + optional per-bullet edits. Actions: accept/reject one, accept-section, reject-section, accept-all, reject-all, reset.
3. **UI** on each aligned bullet (inline diff from the diff issue): Accept / Reject / Edit controls; section-level Accept-all / Reject-all; global Apply bar showing accepted count. Reuse `Button`, `EditableField` (edit-in-place), `Card`, `InlineResult`.
4. **Apply** — pure `applyAcceptedToResume(reconstructed, decisions, edits)` merges accepted proposed bullets into `ReconstructedResume` (replace matched original, insert additions, drop rejected). Target the existing `ReconstructedResume.tsx` / `ReconstructedRole.tsx` surface; no DB.
5. Preserve number-preservation warnings per accepted bullet.

## Dependencies

- Blocked by #209 (inline-diff) — needs `InlineDiff` + `computeTextDiff` to render per-bullet diffs.

## Acceptance

- Each proposed bullet shows an inline diff vs its aligned original and has working Accept / Reject / Edit.
- Section-level accept-all/reject-all and a global Apply work.
- Apply writes accepted bullets into the reconstructed resume; rejected ones are dropped; edits respected.
- Alignment + apply functions unit-tested (M≠N, additions, removals, edits).
- Passes `npm run lint` / style guard; no hardcoded colors; reuses design-system primitives (no hand-rolled buttons).

## References

- Recruidea: `dashboard/hooks/useArtifactReview.ts` (decisions Map, acceptAllHighConfidence, rejectRemaining, applyAccepted, applySuggestionsToResume), `dashboard/components/ArtifactSuggestionReview.tsx`
- ResumeLint targets: `src/components/features/SectionRewrite.tsx`, `src/components/features/ResumeRewriteProposed.tsx`, `src/components/features/ReconstructedResume.tsx`, `src/hooks/useResumeRewrite.ts`

### PR DESCRIPTION
**Pull request by s-annam**
## Summary

Two rewrite-flow features landed on one branch:

- **#210 — Steer the rewrite prompt.** A freeform "Instructions for the rewrite" box plus page-length preset chips (1/2/3 pages) near the rewrite trigger. Instructions are injected as a *system-prompt suffix* (after the existing guardrails, never interleaved) via the new pure `src/lib/webllm/steering.ts`; page targets map to a length budget + recency-weighted older-experience-compression guidance. Blank box + no chip → byte-identical prompt (no behaviour change). New `TextAreaField` design-system primitive; instructions/target persisted in `localStorage`.
- **#211 — Per-bullet review + apply.** Each proposed bullet is aligned to its nearest original (pure `alignBullets`, order-preserving Needleman–Wunsch over word-level Dice similarity), rendered as an inline redline (reusing #209's `InlineDiff`/`computeTextDiff`), and individually Accept / Reject / Edit. Section-level accept-all/reject-all + a global Apply bar. Accepted bullets merge back into the reconstructed résumé via pure `applyAcceptedBullets` (replace / add / new `removeBullet` capability). Cloud persistence punted (paid-tier lane).

Per-bullet apply is wired on the per-role `SectionRewrite` surface (the reconstructed-résumé edit surface that owns the bullet callbacks); the whole-résumé proposed panel keeps its existing inline-diff + Discard.

Resolves #210
Resolves #211

## Test plan

- [x] `npm run typecheck` clean
- [x] `npm run test` green (997/997, 73 files)
- [x] `npm run lint` / style guard clean (DS primitives only, no raw interactive HTML, no hardcoded colors)
- [x] `alignBullets` + `applyAcceptedBullets` unit-tested (M≠N, additions, removals, edits, reorder, duplicates)
- [x] Manually verified in `npm run dev`

### CODE REVIEW

File: src/components/features/ResumeRewrite.tsx [hunk 1]
```diff
@@ -65,12 +82,77 @@
...
+}
+
+/**
+ * Freeform "Instructions for the rewrite" box with page-length preset chips
+ * (#210). One input: the chips prefill editable guidance into the same box and
+ * set the page-length target. Inputs disable while any rewrite is in flight so
+ * the user can't change steering mid-run. Reuses the `TextAreaField` and
+ * `Button` primitives — no raw `<textarea>` / `<button>`.
+ */
+export function RewriteSteeringBox({
```
[the remark this answers is not in the document]
Comment (s-annam) [reply src/components/features/ResumeRewrite.tsx]: Fixed in `526a031` — dropped the `export` on `RewriteSteeringBox`; it's only consumed in-file (ResumeRewrite.tsx:85), so it was never meant to be a module export.

File: src/components/features/RewriteReviewList.tsx [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    ? edited
+    : pair.proposed;
+}
+
+function oldSideText(pair: AlignedPair): string {
+  return pair.kind === "added" ? "" : pair.original;
+}
+
+/** One reviewable bullet: redline + accept/reject + (for non-removals) edit. */
+function BulletReviewRow({
```
[the remark this answers is not in the document]
Comment (s-annam) [reply src/components/features/RewriteReviewList.tsx]: Fixed in `526a031` — `BulletReviewRow` was at 0% coverage (CRAP 72). Added `RewriteReviewList.test.tsx` (jsdom render) exercising all three `AlignedPair` kinds plus accept/reject/edit/bulk controls, dropping CRAP to the function's bare cyclomatic (8).

File: src/lib/rewrite-review/align-bullets.ts [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+ * Backtracking from the bottom-right corner yields the pairs in reading order;
+ * ties favour `matched` over the gaps, then `removed` (original) before
+ * `added` (proposed) so the diff reads original-then-new at a divergence.
+ *
+ * Edge cases:
+ *   - empty `original`  → every proposed is `added`, in order.
+ *   - empty `proposed`  → every original is `removed`, in order.
+ *   - both empty        → `[]`.
+ */
+export function alignBullets(
```
[the remark this answers is not in the document]
Comment (s-annam) [reply src/lib/rewrite-review/align-bullets.ts]: Fixed in `526a031` — extracted `gatedSimMatrix`, `fillDpTable`, `backtrackPairs`, and a shared `diagScore`/`addedPair`/`removedPair` out of `alignBullets`. Cognitive complexity now under the gate (was 31); behavior unchanged (align-bullets.test.ts still green).

**Comment by s-annam** [comment]
image
image

**Review by s-annam — COMMENTED** [review]
## 🤖 Claude review — rewrite steering + per-bullet review (#210, #211)

_Posted by Claude Code on Sri's behalf. Two independent sub-reviewers (React/TSX + TS-lib) audited the full 26-file diff against the checked-out branch; findings triaged below._

### Summary
Branch was BEHIND `main` — merged `origin/main` (clean, no conflicts) and pushed. **Gates green on the merged branch:** typecheck clean, lint clean, **1009 tests pass (75 files)**, build implied by CI `verify`.

The core logic is sound. The Needleman–Wunsch bullet alignment (`align-bullets.ts`) is correct — boundary cases (i=0/j>0, j=0/i>0), the gap-vs-match tie-break (gap score 0 < MATCH_THRESHOLD 0.3), and id-uniqueness via monotone backtracking all check out. The apply index-mapping (`resolveSectionWrites` joining section-relative `originalIndex` → real `BulletObservation.index` via the parallel `obsIndices[]`) is bounds-guarded (`undefined || < 0`) and the arrays are derived from the same `group.bullets` snapshot, so they stay parallel. Steering injection, the review hooks, and the override-apply path are all clean.

**No blocking findings.** Items below are Suggestions/Nits.

### Findings

**[Suggestion] `ResumeRewrite.tsx` — chip switch leaves a stale hint.** Pick "1 page" (prefills the box), then switch to "2 pages" without clearing: the box still reads "...single page" while `pageTarget=2`. Internally inconsistent (box copy vs. length budget). `onChipClick` prefills only when the box is empty by design; consider also replacing the box when switching between non-null presets *if* the current content is a known `PRESET_HINTS` value (never clobber user-typed text).

**[Suggestion] `Dialog.tsx` className contract.** The JSDoc calls `className` "inner panel layout", but it's concatenated onto the `<dialog>` element itself (`cls = ${CHROME} ${className}`, L100/109), not an inner wrapper. The launcher passes `fixed left-1/2 top-1/2 -translate-x-1/2 -translate-y-1/2` for centering. This *works* (positioning utilities, not `display`, so the UA `dialog:not([open]){display:none}` rule is intact and centering is correct) — but the JSDoc and the application site disagree. Either move `className` onto a child `<div>` in the primitive, or fix the JSDoc to say it lands on `<dialog>`. Low priority; don't regress the working centering.

**[Suggestion] `SectionRewrite.tsx` — no stale-source guard on the per-section path.** Unlike the whole-résumé path (`useResumeRewrite`'s `sectionsEqual` guard), a per-section proposal stays open if the user edits a bullet mid-review. Index VALUES stay stable (BulletObservation indices don't shift on text-only edits), so no corruption — but an accepted replacement can silently overwrite the user's own mid-review edit. A one-line comment documenting this trade-off would stop a future maintainer from "fixing" it with a dismiss guard that breaks the flow.

**[Nit] `ResumeRewriteProposed.tsx` L104 / `SectionRewrite.tsx` — `onApply` not `useCallback`.** Benign today (`Button` isn't `React.memo`-wrapped). Wrapping with explicit deps would future-proof and match the pattern already used elsewhere.

**[Nit] `RewriteReviewList.tsx` L81 — rejected state uses `variant="primary"` (amber fill).** Distinguishable from accepted (separate button, "Rejected" vs "Accepted" text), so functional — but an amber fill reads as positive for a reject. A muted/danger treatment would communicate intent better.

**[Nit] `ResumeRewrite.tsx` — "Run rewrite" `aria-label="Run the full-résumé rewrite"` differs from its visible text.** SR users hear a different phrase than sighted users read. Drop the redundant `aria-label` (visible text already describes it) or make them match.

### Fallow (report-only, non-blocking)
Three code-scanning alerts on the pre-merge head; on the fresh CI run:
- `RewriteSteeringBox` "export never imported" — **stale**: no longer exported (local fn now). Auto-clears.
- `BulletReviewRow` CRAP 72 @ 0% coverage — this PR adds `RewriteReviewList.test.tsx` + `ResumeRewriteProposed.test.tsx` that render it; coverage rises, score drops.
- `alignBullets` cognitive complexity 31 (>15) — **real and persistent**. Inherent to Needleman–Wunsch; report-only. Optional: extract the backtracking loop into a helper to bring it under threshold.

### Verdict
**COMMENT** — no blocking issues; logic verified correct and gates green. Two Suggestions worth a quick pass (chip stale-hint, the two doc/comment notes); the rest are Nits. Requesting @Vaishnavi1709 for the approving review.

**Comment by s-annam** [comment]
## 🤖 Claude — review findings addressed (`96c6bb3`)

_Posted by Claude Code on Sri's behalf. All actionable items from the prior review pass are in._

| # | Finding | Resolution |
|---|---------|------------|
| Suggestion | `ResumeRewrite` chip switch left a stale hint (1→2 pages still read "single page") | Fixed — `onChipClick` now refreshes the box when it holds a recognized preset hint; still never clobbers user-typed text |
| Suggestion | `Dialog` `className` JSDoc said "inner panel" but applies to `<dialog>` | Fixed — JSDoc now states classes land on the `<dialog>` element (positioning utils only; documents the closed-state `display` caveat) |
| Suggestion | `SectionRewrite` per-section stale-source guard | No change — the guard already exists (`bulletsEqual` auto-dismiss, L186–192) and the whole-résumé global-apply trade-off is documented in `ResumeRewriteProposed` L90–93 |
| Nit | `onApply` not memoized | Fixed — wrapped in `useCallback` (matches the per-role path) |
| Nit | Reject button amber `primary` fill read as positive | Fixed — neutral-muted treatment when rejected, distinct from the amber accept |
| Nit | "Run rewrite" `aria-label` ≠ visible text | Fixed — dropped the redundant `aria-label` (visible text names it) |

**Gates green on `96c6bb3`:** typecheck clean, lint clean, **1009 tests pass (75 files)**. fallow alerts from the prior run were stale/report-only and auto-cleared.

@Vaishnavi1709 re-requesting your review — these are all low-risk Suggestion/Nit fixes; the core alignment + apply logic is unchanged.

**Review by Vaishnavi1709 — APPROVED** [review]
